### PR TITLE
Get type definition at position

### PR DESCRIPTION
**Pull request by mhegazy**
Implementation for goto type in the services and server layers.

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -987,6 +987,8 @@ module ts {
         findRenameLocations(fileName: string, position: number, findInStrings: boolean, findInComments: boolean): RenameLocation[];
 
         getDefinitionAtPosition(fileName: string, position: number): DefinitionInfo[];
+        getTypeDefinitionAtPosition(fileName: string, position: number): DefinitionInfo[];
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: We should review for a common name with other LS teams.

File: src/harness/harnessLanguageService.ts [hunk 2]
```diff
@@ -333,6 +333,9 @@ module Harness.LanguageService {
         getDefinitionAtPosition(fileName: string, position: number): ts.DefinitionInfo[] {
             return unwrapJSONCallResult(this.shim.getDefinitionAtPosition(fileName, position));
         }
+        getTypeDefinitionAtPosition(fileName: string, position: number): ts.DefinitionInfo[]{
```
Comment (CyrusNajmabadi) [inline src/harness/harnessLanguageService.ts]: What is the difference between getDefinitionAtPositoin and getTypeDefinitionAtPosition?

Comment (mhegazy) [inline src/harness/harnessLanguageService.ts]: One gets you the place where the "symbol" is defined, where the other takes you to the definition location of the "type".

it is a very useful feature when you want to know what is the shape of `value.property`, hovering gives you a type name, but to know the shape of the type you need to first jump to the declaration of `property`, then jump to the declaration of its type. it gets even harder with inferred types, cause you need to go to the definition of the function, or variable that produces the contextual type then go to the type from there.

editors like Eclipse and monaco have had this feature for long time.

File: src/server/protocol.d.ts [hunk 3]
```diff
@@ -126,6 +126,14 @@ declare module ts.server.protocol {
     }
 
     /**
+      * Go to type request; value of command field is
+      * "type". Return response giving the file locations that
```
Comment (DanielRosenwasser) [inline src/server/protocol.d.ts]: That sounds vague enough that I could see it becoming a problem, let's make it "gototype" or something

File: src/server/session.ts [hunk 4]
```diff
@@ -285,7 +286,29 @@ module ts.server {
...
+            }
+
+            return definitions.map(def => ({
+                file: def.fileName,
+                start: compilerService.host.positionToLineOffset(def.fileName, def.textSpan.start),
+                end: compilerService.host.positionToLineOffset(def.fileName, ts.textSpanEnd(def.textSpan))
+            }));
+        }
+
+        getOccurrences(line: number, offset: number, fileName: string): protocol.OccurrencesResponseItem[]{
```
Comment (DanielRosenwasser) [inline src/server/session.ts]: Space before `{`

**Comment by DanielRosenwasser** [comment]
Let's get a test in for primitives: `string`, `boolean`, `number`, `any`, `symbol`.

**Comment by DanielRosenwasser** [comment]
Let's also get a test for toToType on a constructor (i.e. what happens if I have `class C { }; C/*ref*/;` and I request the type of `C`?

**Comment by DanielRosenwasser** [comment]
Oh, and also a test for something with a type literal.

File: src/server/protocol.d.ts (hunk 3 above)
Comment (mhegazy) [inline src/server/protocol.d.ts]: I thought about that... but everything else was short.. 
"gototype", why not "gotodefinition" then
"typedefintion", long, but i like it better.

File: src/harness/harnessLanguageService.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/harness/harnessLanguageService.ts]: That makes sense.  We shoudl drive this through on the Roslyn side as well though.  Definitely open up a bug on them to support this.  That way we can get a command in for VB/C#/TS for this scenario, and the experience can be consistent between all three languages.

File: src/services/shims.ts [hunk 5]
```diff
@@ -589,6 +597,20 @@ module ts {
                 });
         }
 
+        /// GOTO Type
```
Comment (DanielRosenwasser) [inline src/services/shims.ts]: Remove this line

**Comment by DanielRosenwasser** [comment]
Remove the line I mentioned and then :+1:

File: src/services/shims.ts (hunk 5 above)
Comment (mhegazy) [inline src/services/shims.ts]: I added it to match the /// GOTO Definition above

**Pull request merged**